### PR TITLE
Replace type system and parser

### PR DESCRIPTION
**Pull request by pentacular**
This is a fairly radical change, but I think it addresses (at least my) fundamental existing issues. :)
I've replaced the type and parsing system.

The type system is now eagerly resolved and independent of introspection.
It is also has interfaces for resolving argument, parameter, return, property, etc, types.

Parsing is delegated to 'doctrine', which means that types like 'string|function(number,(string|number))' can be parsed.

Only the global jsdoc namespace is supported at this point, but now that the definitions are parsed once, in order, the module namespace, etc, can now be supported.

It is not as clean as I would like, but I think this is a good point to consolidate and consider future directions.

### CODE REVIEW

**Comment by pentacular** [comment]
Will update with array types and some fixes to import, etc.

**Comment by pentacular** [comment]
Arrays, imports, etc.

Seems to more or less work on a real module now.

**Comment by pentacular** [comment]
A few things left to sort out

/** @typedef {boolean} Foo */
/** @typedef {boolean} Bar */

Currently Foo and Bar are equivalent aliases to boolean.
But they should really be mutually incompatible subtypes of boolean.

Foo should be usable where boolean is required, but not where Bar is required.

File: package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
   "main": "src/index.js",
   "dependencies": {
     "comment-parser": "^0.7.2",
+    "doctrine": "^3.0.0",
```
Comment (orzechowskid) [inline package.json:14]: this looks much better than the other JSDoc parsing library we were looking at.  good find.

File: src/rules/tests/assignment-types-must-match.test.js [hunk 2]
```diff
@@ -197,8 +197,8 @@ var x = foo.bar();
         });
 
         it(`should not show a message`, function() {
```
Comment (orzechowskid) [inline src/rules/tests/assignment-types-must-match.test.js:199]: the `it` description for this test is no longer valid now that it does display a message.

File: src/rules/tests/assignment-types-must-match.test.js [hunk 3]
```diff
@@ -283,9 +283,9 @@ var x = { data: { name: 'alice', value: undefined }, department: 'finance' };
             result = await doTest(source, lintOptions);
         });
 
-        it(`should show a message`, function() {
+        it(`should not show a message`, function() {
```
Comment (orzechowskid) [inline src/rules/tests/assignment-types-must-match.test.js:286]: same here

File: src/utils.js [hunk 4]
```diff
@@ -277,70 +83,87 @@ function visitFile(context) {
...
 
 function getContextForFile(fsPath, currentContext) {
+    if (!fsPath.endsWith('.js')) {
+      fsPath = `${fsPath}.js`;
+    }
+
+    const resolvedPath = resolve(fsPath, currentContext);
     const newContext = {};
 
+    // CHECK: Does this do something different to Object.assign()?
```
Comment (orzechowskid) [inline src/utils.js:106]: `Object.assign` only copies an object's own properties; `for..in` copies inherited properties too.  The context object we get from eslint must be subclassed from something else (I didn't dig into it too much)

File: src/rules/tests/assignment-types-must-match.test.js (hunk 3 above)
Comment (pentacular) [reply src/rules/tests/assignment-types-must-match.test.js:286]: done

File: src/rules/tests/assignment-types-must-match.test.js (hunk 2 above)
Comment (pentacular) [reply src/rules/tests/assignment-types-must-match.test.js:199]: done

File: package.json (hunk 1 above)
Comment (pentacular) [reply package.json:14]: Yeah, the only concern is that it is frozen and no-longer maintained.

File: src/utils.js (hunk 4 above)
Comment (pentacular) [reply src/utils.js:106]: Thanks

**Comment by pentacular** [comment]
I think we should be able to use the type resolution system outside of eslint.
e.g., to be able to do is to generate guard functions from function types for dynamic type checking.

So I've moved the AST based type derivations across, leaving the identifier based lookup in util.js.

I'd like to move the type logic to another (preferably es6 style) module -- what do you think?

**Comment by orzechowskid** [comment]
I agree - thanks to all the work you've done to help make this an actual useful library instead of just a project that worked for the one thing I used it on :) , the type-resolution code is useful enough to be its own module, or even its own library if that's what you were thinking.

I used CJS-style files instead of ES6-style modules since I wanted compatibility with as many versions of node as possible and didn't want to have to deal with a transpiler.  but I suppose it's not hard to use modern syntax and transpile to CJS only as opposed to UMD (and it would be nice to have object spread syntax again...).

let me finish digesting this PR first, then I'll probably merge this and release a new version.  that might be a good time to introduce any tooling changes we want.

**Comment by pentacular** [comment]
Sounds good.

I've held off on adding test cases for Type for now -- but it should be tested also at that level.

**Pull request merged**